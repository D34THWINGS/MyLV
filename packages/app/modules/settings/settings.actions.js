import { disableDesktopNotifications, enableDesktopNotifications } from './push.service'
import { fetchWithAuth, fetchUserData } from '../auth/auth.actions'

export const TOGGLE_PROCESS_REMINDER = 'TOGGLE_PROCESS_REMINDER'
export const toggleProcessReminder = (toggle = false) => ({ type: TOGGLE_PROCESS_REMINDER, payload: toggle })

export const TOGGLE_PROOF_OF_TRANSPORT_DIALOG = 'TOGGLE_PROOF_OF_TRANSPORT_DIALOG'
export const toggleProofOfTransportDialog = (toggle = false) => ({
  type: TOGGLE_PROOF_OF_TRANSPORT_DIALOG,
  payload: toggle,
})

export const TOGGLE_PUSH_NOTIFICATION_SNACK = 'TOGGLE_PUSH_NOTIFICATION_SNACK'
export const togglePushNotificationSnack = (toggle = false) => ({
  type: TOGGLE_PUSH_NOTIFICATION_SNACK,
  payload: toggle,
})

export const TOGGLE_PUSH_NOTIFICATIONS = 'TOGGLE_PUSH_NOTIFICATIONS'
export const togglePushNotifications = () => (dispatch, getState) => {
  const state = getState()
  const { subscriptionId, desktopNotificationsEnabled: isSubscribed } = state.settings

  if (isSubscribed) {
    return disableDesktopNotifications()
      .then(() => {
        if (!subscriptionId) {
          return Promise.resolve()
        }
        return dispatch(fetchWithAuth(`/api/subscriptions/${subscriptionId}`, { method: 'DELETE' }))
      })
      .then(() => dispatch({ type: TOGGLE_PUSH_NOTIFICATIONS, payload: { enabled: false, id: null } }))
  }

  return enableDesktopNotifications()
    .then(subscription => dispatch(fetchWithAuth('/api/subscriptions', { method: 'POST', body: subscription })))
    .then(({ id }) => dispatch({ type: TOGGLE_PUSH_NOTIFICATIONS, payload: { enabled: true, id } }))
    .catch(e => console.error(e))
}

export const DESKTOP_NOTIFICATIONS_INSTALLED = 'DESKTOP_NOTIFICATIONS_INSTALLED'
export const desktopNotificationsInstalled = isSubscribed => ({
  type: DESKTOP_NOTIFICATIONS_INSTALLED,
  payload: { isSubscribed },
})

export const TOGGLE_HOLIDAYS_DISCLAIMER = 'TOGGLE_HOLIDAYS_DISCLAIMER'
export const toggleHolidaysDisclaimer = (value = false) => ({
  type: TOGGLE_HOLIDAYS_DISCLAIMER,
  payload: value,
})

export const TOGGLE_TUTORIALS = 'TOGGLE_TUTORIALS'
export const toggleTutorials = (value = false) => ({
  type: TOGGLE_TUTORIALS,
  payload: value,
})

export const SAVE_PREFERENCES_START = 'SAVE_PREFERENCES_START'
export const SAVE_PREFERENCES_SUCCESS = 'SAVE_PREFERENCES_SUCCESS'
export const SAVE_PREFERENCES_ERROR = 'SAVE_PREFERENCES_ERROR'
export const savePreferences = changes => async (dispatch) => {
  dispatch({ type: SAVE_PREFERENCES_START })

  try {
    const profile = await dispatch(fetchWithAuth('/api/profile/me', { method: 'PUT', body: changes }))

    // Reload user data with latest proof of transport
    if (changes.hasProofOfTransport) {
      await dispatch(fetchUserData())
    }

    dispatch({ type: SAVE_PREFERENCES_SUCCESS, payload: profile })
  } catch (e) {
    dispatch({ type: SAVE_PREFERENCES_ERROR, payload: e })
  }
}
