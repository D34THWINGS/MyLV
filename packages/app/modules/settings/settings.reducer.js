import { REHYDRATE } from 'redux-persist/constants'
import {
  TOGGLE_PROCESS_REMINDER,
  DESKTOP_NOTIFICATIONS_INSTALLED,
  TOGGLE_PUSH_NOTIFICATIONS,
  TOGGLE_PUSH_NOTIFICATION_SNACK,
  TOGGLE_HOLIDAYS_DISCLAIMER,
  TOGGLE_TUTORIALS,
  SAVE_PREFERENCES_SUCCESS,
} from './settings.actions'
import { RECEIVE_USER_DATA } from '../auth/auth.actions'

const initialState = {
  shouldRemindProcess: true,
  shouldDisplayHolidaysDisclaimer: true,
  shouldDisplayProofOfTransportDialog: false,
  shouldDisplayPushNotificationSnack: true,
  desktopNotificationsEnabled: false,
  desktopNotificationsInstalled: false,
  subscriptionId: null,
  rehydrated: false,
}

export default (state = initialState, { type, payload }) => {
  switch (type) {
    case TOGGLE_PROCESS_REMINDER:
      return {
        ...state,
        shouldRemindProcess: payload,
      }
    case SAVE_PREFERENCES_SUCCESS:
      return {
        ...state,
        shouldDisplayProofOfTransportDialog: payload.hasProofOfTransport,
      }
    case DESKTOP_NOTIFICATIONS_INSTALLED:
      return {
        ...state,
        desktopNotificationsInstalled: true,
        desktopNotificationsEnabled: payload.isSubscribed,
      }
    case TOGGLE_PUSH_NOTIFICATIONS:
      return {
        ...state,
        desktopNotificationsEnabled: payload.enabled,
        subscriptionId: payload.id,
        shouldDisplayPushNotificationSnack: false,
      }
    case TOGGLE_PUSH_NOTIFICATION_SNACK:
      return {
        ...state,
        shouldDisplayPushNotificationSnack: payload,
      }
    case TOGGLE_HOLIDAYS_DISCLAIMER:
      return {
        ...state,
        shouldDisplayHolidaysDisclaimer: payload,
      }
    case TOGGLE_TUTORIALS:
      return {
        ...state,
        shouldDisplayHolidaysDisclaimer: payload,
        shouldRemindProcess: payload,
      }
    case RECEIVE_USER_DATA:
      return {
        ...state,
        shouldDisplayProofOfTransportDialog: payload.profile.hasProofOfTransport,
      }
    case REHYDRATE:
      return {
        ...state,
        ...payload.settings,
        shouldDisplayProofOfTransportDialog: state.shouldDisplayProofOfTransportDialog,
        desktopNotificationsEnabled: state.desktopNotificationsEnabled,
        desktopNotificationsInstalled: state.desktopNotificationsInstalled,
        rehydrated: true,
      }
    default:
      return state
  }
}
