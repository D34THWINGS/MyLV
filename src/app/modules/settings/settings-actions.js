export const STOP_PROCESS_REMINDER = 'STOP_PROCESS_REMINDER'
export const stopProcessReminder = () => ({ type: STOP_PROCESS_REMINDER })

export const DISABLE_PROOF_OF_TRANSPORT_DIALOG = 'DISABLE_PROOF_OF_TRANSPORT_DIALOG'
export const disableProofOfTransportDialog = () => ({ type: DISABLE_PROOF_OF_TRANSPORT_DIALOG })
