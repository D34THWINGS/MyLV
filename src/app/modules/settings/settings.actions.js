import { disableDesktopNotifications, enableDesktopNotifications } from './push.service'
import { fetchWithAuth } from '../auth/auth.actions'

export const TOGGLE_PROCESS_REMINDER = 'TOGGLE_PROCESS_REMINDER'
export const toggleProcessReminder = (toggle = false) => ({ type: TOGGLE_PROCESS_REMINDER, payload: toggle })

export const TOGGLE_PROOF_OF_TRANSPORT_DIALOG = 'TOGGLE_PROOF_OF_TRANSPORT_DIALOG'
export const toggleProofOfTransportDialog = (toggle = false) => ({
  type: TOGGLE_PROOF_OF_TRANSPORT_DIALOG,
  payload: toggle,
})

export const TOGGLE_PUSH_NOTIFICATION_SNACK = 'TOGGLE_PUSH_NOTIFICATION_SNACK'
export const togglePushNotificationSnack = (toggle = false) => ({
  type: TOGGLE_PUSH_NOTIFICATION_SNACK,
  payload: toggle,
})

export const TOGGLE_PUSH_NOTIFICATIONS = 'TOGGLE_PUSH_NOTIFICATIONS'
export const togglePushNotifications = () => (dispatch, getState) => {
  const state = getState()
  const { subscriptionId, desktopNotificationsEnabled: isSubscribed } = state.settings

  if (isSubscribed) {
    return disableDesktopNotifications()
      .then(() => {
        if (!subscriptionId) {
          return Promise.resolve()
        }
        return dispatch(fetchWithAuth(`/api/subscriptions/${subscriptionId}`, { method: 'DELETE' }))
      })
      .then(() => dispatch({ type: TOGGLE_PUSH_NOTIFICATIONS, payload: { enabled: false, id: null } }))
  }

  return enableDesktopNotifications()
    .then(subscription => dispatch(fetchWithAuth('/api/subscriptions', { method: 'POST', body: subscription })))
    .then(({ id }) => dispatch({ type: TOGGLE_PUSH_NOTIFICATIONS, payload: { enabled: true, id } }))
    .catch(e => console.error(e))
}

export const DESKTOP_NOTIFICATIONS_INSTALLED = 'DESKTOP_NOTIFICATIONS_INSTALLED'
export const desktopNotificationsInstalled = isSubscribed => ({
  type: DESKTOP_NOTIFICATIONS_INSTALLED,
  payload: { isSubscribed },
})
