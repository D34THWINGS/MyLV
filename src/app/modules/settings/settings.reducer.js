import { REHYDRATE } from 'redux-persist/constants'
import {
  TOGGLE_PROCESS_REMINDER,
  TOGGLE_PROOF_OF_TRANSPORT_DIALOG,
  DESKTOP_NOTIFICATIONS_INSTALLED,
  TOGGLE_PUSH_NOTIFICATIONS,
  TOGGLE_PUSH_NOTIFICATION_SNACK,
  TOGGLE_HOLIDAYS_DISCLAIMER,
  TOGGLE_TUTORIALS,
} from './settings.actions'

const initialState = {
  shouldRemindProcess: true,
  shouldDisplayHolidaysDisclaimer: true,
  shouldDisplayProofOfTransportDialog: true,
  shouldDisplayPushNotificationSnack: true,
  desktopNotificationsEnabled: false,
  desktopNotificationsInstalled: false,
  subscriptionId: null,
  rehydrated: false,
}

export default (state = initialState, { type, payload }) => {
  switch (type) {
    case TOGGLE_PROCESS_REMINDER:
      return {
        ...state,
        shouldRemindProcess: payload,
      }
    case TOGGLE_PROOF_OF_TRANSPORT_DIALOG:
      return {
        ...state,
        shouldDisplayProofOfTransportDialog: payload,
      }
    case DESKTOP_NOTIFICATIONS_INSTALLED:
      return {
        ...state,
        desktopNotificationsInstalled: true,
        desktopNotificationsEnabled: payload.isSubscribed,
      }
    case TOGGLE_PUSH_NOTIFICATIONS:
      return {
        ...state,
        desktopNotificationsEnabled: payload.enabled,
        subscriptionId: payload.id,
        shouldDisplayPushNotificationSnack: false,
      }
    case TOGGLE_PUSH_NOTIFICATION_SNACK:
      return {
        ...state,
        shouldDisplayPushNotificationSnack: payload,
      }
    case TOGGLE_HOLIDAYS_DISCLAIMER:
      return {
        ...state,
        shouldDisplayHolidaysDisclaimer: payload,
      }
    case TOGGLE_TUTORIALS:
      return {
        ...state,
        shouldDisplayHolidaysDisclaimer: payload,
        shouldRemindProcess: payload,
      }
    case REHYDRATE:
      return {
        ...state,
        ...payload.settings,
        desktopNotificationsEnabled: state.desktopNotificationsEnabled,
        desktopNotificationsInstalled: state.desktopNotificationsInstalled,
        rehydrated: true,
      }
    default:
      return state
  }
}
