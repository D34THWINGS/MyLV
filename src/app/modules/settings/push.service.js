import { featureFlipping, push } from '../../config'
import { desktopNotificationsInstalled } from './settings.actions'

let pushWorker

function urlB64ToUint8Array(base64String) {
  const padding = '='.repeat((4 - (base64String.length % 4)) % 4)
  const base64 = (base64String + padding)
    .replace(/-/g, '+')
    .replace(/_/g, '/')

  const rawData = window.atob(base64)
  const outputArray = new Uint8Array(rawData.length)

  for (let i = 0; i < rawData.length; i += 1) {
    outputArray[i] = rawData.charCodeAt(i)
  }
  return outputArray
}

export const installPushNotifications = async (store) => {
  if (!navigator.serviceWorker || !featureFlipping.pushNotifications) {
    return
  }

  try {
    pushWorker = await navigator.serviceWorker.getRegistration('/')
    const subscription = await pushWorker.pushManager.getSubscription()
    const isSubscribed = !(subscription === null)
    store.dispatch(desktopNotificationsInstalled(isSubscribed))
  } catch (e) {
    console.error(e)
  }
}

export const enableDesktopNotifications = () => {
  if (!pushWorker) {
    return Promise.reject(new Error('Desktop notification not installed'))
  }

  return pushWorker.pushManager.subscribe({
    userVisibleOnly: true,
    applicationServerKey: urlB64ToUint8Array(push.publicKey),
  })
}

export const disableDesktopNotifications = () => {
  if (!pushWorker) {
    return Promise.reject(new Error('Desktop notification not installed'))
  }

  return pushWorker.pushManager.getSubscription()
    .then((subscription) => {
      if (subscription) {
        return subscription.unsubscribe()
      }

      return Promise.resolve()
    })
}
