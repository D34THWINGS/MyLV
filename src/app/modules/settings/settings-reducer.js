import { STOP_PROCESS_REMINDER, DISABLE_PROOF_OF_TRANSPORT_DIALOG } from './settings-actions'

const initialState = {
  shouldRemindProcess: true,
  shouldDisplayProofOfTransportDialog: true,
}

export default (state = initialState, action) => {
  switch (action.type) {
    case STOP_PROCESS_REMINDER:
      return {
        ...state,
        shouldRemindProcess: false,
      }
    case DISABLE_PROOF_OF_TRANSPORT_DIALOG:
      return {
        ...state,
        shouldDisplayProofOfTransportDialog: false,
      }
    default:
      return state
  }
}
