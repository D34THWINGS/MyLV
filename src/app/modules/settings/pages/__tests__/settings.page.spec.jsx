import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'

import { SettingsPage } from '../settings.page'

jest.unmock('../settings.page')

describe('SettingsPage', () => {
  let props

  const getWrapper = () => shallow(<SettingsPage {...props} />)

  beforeEach(() => {
    props = {
      togglePushNotifications: jest.fn(),
      toggleProofOfTransportDialog: jest.fn(),
      toggleProcessReminder: jest.fn(),
      settings: {
        desktopNotificationsInstalled: false,
        desktopNotificationsEnabled: false,
        shouldDisplayProofOfTransportDialog: false,
        shouldRemindProcess: false,
      },
    }
  })

  it('should render with all switches off', () => {
    // When
    const wrapper = getWrapper()

    // Then
    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('should render with all switches on', () => {
    // Given
    props.settings = {
      desktopNotificationsInstalled: true,
      desktopNotificationsEnabled: true,
      shouldDisplayProofOfTransportDialog: true,
      shouldRemindProcess: true,
    }

    // When
    const wrapper = getWrapper()

    // Then
    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })
})
