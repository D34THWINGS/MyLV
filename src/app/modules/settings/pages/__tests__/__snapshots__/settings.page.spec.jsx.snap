// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsPage should render with all switches off 1`] = `
<Card
  raised={false}
>
  <WithStyles(CardContent)>
    <WithStyles(Typography)
      component="h2"
      gutterBottom={true}
      variant="headline"
    >
      Paramètres
    </WithStyles(Typography)>
    <WithStyles(List)>
      <WithStyles(ListItem)>
        <WithStyles(ListItemText)
          primary="Notifications bureau de rappel pour le CRA"
        />
        <WithStyles(ListItemSecondaryAction)>
          <WithStyles(Switch)
            checked={false}
            disabled={true}
            onChange={[MockFunction]}
          />
        </WithStyles(ListItemSecondaryAction)>
      </WithStyles(ListItem)>
      <WithStyles(ListItem)>
        <WithStyles(ListItemText)
          primary="Popup rappel titre de transport"
        />
        <WithStyles(ListItemSecondaryAction)>
          <WithStyles(Switch)
            checked={false}
            onChange={[MockFunction]}
          />
        </WithStyles(ListItemSecondaryAction)>
      </WithStyles(ListItem)>
      <WithStyles(ListItem)>
        <WithStyles(ListItemText)
          primary="Afficher les tutoriels"
        />
        <WithStyles(ListItemSecondaryAction)>
          <WithStyles(Switch)
            checked={false}
            onChange={[MockFunction]}
          />
        </WithStyles(ListItemSecondaryAction)>
      </WithStyles(ListItem)>
    </WithStyles(List)>
  </WithStyles(CardContent)>
</Card>
`;

exports[`SettingsPage should render with all switches on 1`] = `
<Card
  raised={false}
>
  <WithStyles(CardContent)>
    <WithStyles(Typography)
      component="h2"
      gutterBottom={true}
      variant="headline"
    >
      Paramètres
    </WithStyles(Typography)>
    <WithStyles(List)>
      <WithStyles(ListItem)>
        <WithStyles(ListItemText)
          primary="Notifications bureau de rappel pour le CRA"
        />
        <WithStyles(ListItemSecondaryAction)>
          <WithStyles(Switch)
            checked={true}
            disabled={false}
            onChange={[MockFunction]}
          />
        </WithStyles(ListItemSecondaryAction)>
      </WithStyles(ListItem)>
      <WithStyles(ListItem)>
        <WithStyles(ListItemText)
          primary="Popup rappel titre de transport"
        />
        <WithStyles(ListItemSecondaryAction)>
          <WithStyles(Switch)
            checked={true}
            onChange={[MockFunction]}
          />
        </WithStyles(ListItemSecondaryAction)>
      </WithStyles(ListItem)>
      <WithStyles(ListItem)>
        <WithStyles(ListItemText)
          primary="Afficher les tutoriels"
        />
        <WithStyles(ListItemSecondaryAction)>
          <WithStyles(Switch)
            checked={true}
            onChange={[MockFunction]}
          />
        </WithStyles(ListItemSecondaryAction)>
      </WithStyles(ListItem)>
    </WithStyles(List)>
  </WithStyles(CardContent)>
</Card>
`;
